pkg/crd/flatten.go

```diff
@@ -26,47 +27,176 @@ import (
 	"sigs.k8s.io/controller-tools/pkg/loader"
 )
 
+// ErrorRecorder knows how to record errors.  It wraps the part of
+// pkg/loader.Package that we need to record errors in places were it might not
+// make sense to have a loader.Package
+type ErrorRecorder interface {
+	// AddError records that the given error occurred.
+	// See the documentation on loader.Package.AddError for more information.
+	AddError(error)
+}
+
+// isOrNil checks if val is nil if val is of a nillable type, otherwise,
+// it comparse val to valuInt (which should probably be the zero value).
+func isOrNil(val reflect.Value, valInt interface{}, zeroInt interface{}) bool {
```

Comment:
why the Int postfix?

Reply:
A better name would be isZeroOrNil. But also that is actually wrong: we never compare with zero in case of nillable types.

Reply:
why not make a `isZero` func? nil is the zero value of nillable types.

Reply:
there's an issue with comparibility -- I original was just comparing against the zero type, but encountered some issues with non-comparibility -- reflect is a bit weird in this regard, although it's possible I made a mistake.

> why the Int postfix

`Int` is the `interface{}` form, `val` is the `reflect.Value` form.

Reply:
we can probably get away with just re-calling `Interface` here, though -- it'll be a bit of a performance hit, but it's prob not too big a deal.  Can profile after writing tests.